### PR TITLE
Midi 109/data

### PR DESCRIPTION
**Pull request by WojciechMat**
### Quantizing continuous start time
Quantization can now be performed on start instead of dstart values. 
The code below initializes a quantizer that uses 650 start bins to quantize 20-second piece of music. Each bin is 0.032 seconds long.
```python3
quantizer = MidiATQuantizer(
    n_duration_bins=3,
    n_velocity_bins=3,
    n_start_bins=20,
    sequence_duration=20.0, 
)
```
Using 650 start bins produces still very good-sounding pieces.
When using only *20* start_bins pianoroll of the sequence with applied quantization looks like this:
image

What worries me  is that we do not have a fixed number of notes pe sequence - perhabs we will need to bring back a padding token.

### Multiple tokens per note

The code below creates a `MyTokenizedMidiDataset` with a tokenizer that uses multiple tokens per note.
Here the `dataset` is a dataset build with MidiATQuantizer.

```python3
keys = ["pitch", "start_bin", "duration_bin", "velocity_bin"]
src_encoder = MultiVelocityEncoder(cfg.quantization, keys=keys, )
test_dataset = MyTokenizedMidiDataset(
    dataset=dataset,
    dataset_cfg=cfg,
    encoder=src_encoder,
)
```
#### Example token sequence
Tokens are in order:
"pitch", "start-duration", "velocity", ...
```
src tokens:  ['<CLS>', '69', '0-2', '0', '64', '0-2', '0', '58', '0-2', '0', '76', '1-1', '0', '78', '1-1', '0', '81', '1-1', '0', '84', '1-2', '1', '72', '2-2', '0', '66', '2-2', '0', '73', '2-2', '0', '76', '3-1', '0', '78', '3-0', '1', '81', '3-0', '1', '84', '4-0', '1', '72', '4-0', '0', '78', '4-2', '0', '87', '4-2', '0', '66', '6-2', '0', '55', '6-2', '0', '61', '6-2', '0', '73', '7-1', '0', '75', '7-1', '1', '78', '7-0', '0', '81', '7-2', '1', '69', '7-2', '0', '63', '8-2', '0', '70', '8-2', '0', '73', '9-2', '0', '75', '9-2', '0', '78', '9-2', '0', '81', '9-1', '0', '69', '10-1', '0', '75', '10-2', '0', '84', '10-2', '0', '54', '12-0', '0', '64', '12-2', '0', '73', '12-2', '0', '70', '12-2', '0', '76', '12-1', '1', '78', '12-1', '0', '81', '12-1', '0', '84', '13-1', '0', '54', '14-1', '0', '64', '14-1', '0', '73', '14-1', '1', '76', '14-1', '1', '70', '14-1', '1', '78', '14-0', '0', '81', '14-0', '0', '84', '14-0', '1', '60', '15-1', '0', '74', '15-1', '1', '70', '15-1', '0', '76', '15-1', '1', '78', '15-0', '1', '82', '15-0', '1', '84', '15-0', '1', '66', '15-1', '1', '73', '15-1', '1', '76', '15-1', '1', '78', '15-1', '1', '81', '15-0', '2', '84', '15-0', '2', '72', '16-1', '2', '74', '16-1', '2', '76', '16-1', '2', '78', '16-1', '2', '82', '16-1', '2', '84', '16-1', '2', '78', '16-1', '2', '73', '16-2', '2', '76', '17-2', '1', '79', '17-2', '2', '81', '17-2', '2', '85', '17-1', '2', '84', '17-1', '1', '88', '17-1', '2', '91', '17-1', '1', '93', '17-2', '2', '96', '17-2', '2', '91', '17-2', '1', '88', '17-2', '2', '85', '17-2', '2', '84', '17-2', '1', '81', '17-2', '2', '76', '17-2', '1', '79', '17-2', '0', '73', '17-2', '1', '69', '18-2', '1']
```
### Tokenization review dashboard
You can check out tokenization methods and read about MultiTokEncoders by running
```shell
PYTHONPATH=. streamlit run --server.port 4466 dashboard/tokenization.py
```

### Padding
Padding is added with collate function in the dataloader.
```python
def collate_fn(batch):
    return collate(batch, max_padding=180, pad_idx=1)

train_dataloader = DataLoader(
    train_dataset,
    batch_size=8,
    shuffle=True,
    collate_fn=collate_fn,
)
```
### Training demo
You can run training demo - it will start training for one epoch but it would take about 30 hours to finish - it is only to confirm that passing data to the model in current form is possible.
```shell
python train.py --config-name T5start
```

### CODE REVIEW

**Comment by WojciechMat** [comment]
It turns out 20s of music can contain up to 628 notes in maestro-v1-sustain test split.

Figure_1

**Comment by roszcz** [comment]
> It turns out 20s of music can contain up to 628 notes in maestro-v1-sustain test split.
> 
> Figure_1

It's a good practice to always include axis labels with units when presenting any charts :)

**Comment by WojciechMat** [comment]
I have cleaned it a little bit, decided to only use MultiTok encoders for now, created a demo training script.

Would you like I add a classic tokenization logic (one token per one note) as well?

**Comment by roszcz** [comment]
Looks great, but I'm not able to understand precisely what's going on in terms of quantization and encoding - can you document the tokenization strategies that you implemented? You can describe that in the PR, create a separate markdown file, prepare a streamlit (or gradio) dashboard, or write a blogpost - whichever you think will work best :)

Also, when running the sample training I'm getting an error after a couple of steps:

```
Traceback (most recent call last):
  File "/home/hagrid/workson/midi-hf-transformers/train.py", line 102, in main
    for batch in pbar:
  File "/home/hagrid/.virtualenvs/midi-hf-transformers/lib/python3.10/site-packages/tqdm/std.py", line 1178, in __iter__
    for obj in iterable:
  File "/home/hagrid/.virtualenvs/midi-hf-transformers/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 633, in __next__
    data = self._next_data()
  File "/home/hagrid/.virtualenvs/midi-hf-transformers/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 677, in _next_data
    data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
  File "/home/hagrid/.virtualenvs/midi-hf-transformers/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 51, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/hagrid/.virtualenvs/midi-hf-transformers/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 51, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/hagrid/workson/midi-hf-transformers/data/dataset.py", line 185, in __getitem__
    target_tokens_ids = self.encoder.encode_tgt(record)
  File "/home/hagrid/workson/midi-hf-transformers/data/tokenizer.py", line 43, in encode_tgt
    token_ids = [self.token_to_id[token] for token in tokens]
  File "/home/hagrid/workson/midi-hf-transformers/data/tokenizer.py", line 43, in <listcomp>
    token_ids = [self.token_to_id[token] for token in tokens]
KeyError: '400s'
```

File: data/tokenizer.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+    def tokenize_src(self, record: dict) -> list[str]:
+        tokens = []
+        n_samples = len(record[self.keys[0]])
+        for idx in range(n_samples):
+            # append p and v so that there are no duplicate tokens
+            pitch_token = f"{record[self.keys[0]][idx]:0.0f}p"
+            time_token = f"{record[self.keys[1]][idx]}-{record[self.keys[2]][idx]:0.0f}"
+            velocity_token = f"{record[self.keys[3]][idx]:0.0f}v"
+            tokens += [pitch_token, time_token, velocity_token]
```
Comment (roszcz) [inline data/tokenizer.py]:
```suggestion
            pitch_token = f"PITCH-{record[self.keys[0]][idx]:0.0f}"
            time_token = f"TIME-{record[self.keys[1]][idx]}-{record[self.keys[2]][idx]:0.0f}"
            velocity_token = f"VEL-{record[self.keys[3]][idx]:0.0f}"
            tokens += [pitch_token, time_token, velocity_token]
```

I think there's no technical reason to keep those short, and I suspect we'll have to look at tokens occasionally, so it's better to make them more explicit.

File: data/tokenizer.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    def __init__(self, quantization_cfg: DictConfig, keys: list[str] = None, tgt_bins: int = 650):
+        super().__init__()
+        self.quantization_cfg = quantization_cfg
+        self.tgt_bins = tgt_bins
+        if keys is None:
+            self.keys = ["pitch", "dstart_bin", "duration_bin", "velocity_bin"]
+        else:
+            self.keys = keys
+
+        self.specials = ["<CLS>", "<pad>"]
```
Comment (roszcz) [inline data/tokenizer.py]:
```suggestion
        self.specials = ["<CLS>", "<PAD>"]
```

Unless there is a reason for this lack of lower/upper case consistency :)

File: data/tokenizer.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+            tokens = tokens[1:]
+
+        for idx, token in enumerate(tokens):
+            if token == "<pad>":
+                break
+            if idx % 3 == 0:
+                # pitch token
+                buff.clear()
+                # [:-1] to clear p letter from a token
+                buff.append(eval(token[:-1]))
```
Comment (roszcz) [inline data/tokenizer.py]:
```suggestion
                buff.append(int(token[:-1]))
```

I'm not sure if that's what's happening here, but using eval for string -> number conversion is not recommended, and you should use float() or int() instead

**Comment by WojciechMat** [comment]
last commit should have read "np.digitize" :)
I fixed the bug with demo training and implemented your suggestions, tokenization explanatory dashboard is coming soon.

**Comment by WojciechMat** [comment]
I have added a tokenization dashboard, perhabs it will make some thing cleaner, I might also add some code snippets later.

Run it with
```shell
PYTHONPATH=. streamlit run --server.port 4466 dashboard/tokenization.py
```

**Review by roszcz — APPROVED** [review]

**Pull request merged**